lib/hooks.js

```diff
@@ -66,39 +67,40 @@ const getProxiedHooks = hookType =>
     : [hookType]
 ;
 
-function getHooks(hooked, hookType) {
-  return (hooked.options.hooks || {})[hookType] || [];
-}
-
-const Hooks = {
+class Hooks {
+  _getHooks(hookType) {
+    return this.hooks[hookType] || [];
+  }
   /**
    * Process user supplied hooks definition
    *
    * @param {object} hooks hooks definition
-   *
+   * @param {Hooks} parent
    * @private
-   * @memberof Sequelize
-   * @memberof Sequelize.Model
    */
-  _setupHooks(hooks) {
-    this.options.hooks = {};
-    _.map(hooks || {}, (hooksArray, hookName) => {
+  constructor(hooks, parent) {
+    this.hooks = {};
```

Comment:
This seems like would be a good use-case for a `Map` 😁 But it's not really the scope of this PR so it's not blocking

Reply:
Ideally maps would be used *everywhere* instead of objects, prototype pollution is becoming more and more of an issue.

Reply:
Absolutely. Also, it has a performance benefit to it.

Reply:
Maps are faster than objects?
I'll need some numbers 😛

Reply:
https://jsperf.com/map-vs-object-sequelize/1 :)

Reply:
@eseliger I was interested and gave it a test. Maps have a faster delete, but if just doing a set, then objects are faster.
https://jsperf.com/sequelize-map-vs-obj-2/1

Reply:
Interesting :)
I also added reading the values in again, seems to still be faster (for those who are interested)
https://jsperf.com/sequelize-map-vs-obj-2/3

Reply:
Interesting, in my yesterday benchmarks Maps were universally faster.

Reply:
What operation were you looking at?